Desktop session recording improvements

zmb3 opened the pull request:
This commit introduces 2 changes:

1. Previously, we leveraged a `TeeStreamer` which would ensure that the audit events we emitted would go to both the session recording (so long as the event type passed a simple check) and Teleport's audit log. This has been removed, and we now leverage the stream writer for session recordings, and the service's configured emitter for events that are destined for the audit log. This is a bit more efficient, because TeeStreamer would perform two EmitAuditEvent calls sequentially, and for desktop access we usually only need one.
2. Leverage a discard stream when session recording is disabled. Prior to this, disabling session recordings would prevent them from being uploaded, but wouldn't prevent them from being written to disk, so users would incur a performance penalty (and fill up their disk) even when session recording is disabled.

Updates #16773

zmb3 committed:
Desktop session recording improvements

This commit introduces 2 changes:

1. Previously, we leveraged a `TeeStreamer` which would ensure that
   the audit events we emitted would go to both the session recording
   (so long as the event type passed a simple check) and Teleport's
   audit log. This has been removed, and we now leverage the stream
   writer for session recordings, and the service's configured emitter
   for events that are destined for the audit log. This is a bit more
   efficient, because TeeStreamer would perform two EmitAuditEvent
   calls sequentially, and for desktop access we usually only need one.
2. Leverage a discard stream when session recording is disabled.
   Prior to this, disabling session recordings would prevent them
   from being uploaded, but wouldn't prevent them from being written
   to disk, so users would incur a performance penalty (and fill up
   their disk) even when session recording is disabled.

Updates #16773

ibeckermayer commented on lib/srv/desktop/windows_server.go:
```diff
@@ -428,7 +435,7 @@ func (s *WindowsService) newStreamer(ctx context.Context, recConfig types.Sessio
 		return nil, trace.Wrap(err)
 	}
 
-	return libevents.NewTeeStreamer(fileStreamer, s.cfg.Emitter), nil
```
I'm confused -- does this mean we never needed to send recording events to the `s.cfg.Emitter` in `async` mode? We can just write them to disk, and they will get uploaded later by some other process?

Related comment regarding the overall changes: my understanding is that session recording events and audit events are going to the same place. If that's the case, then it seems that we shouldn't need a separate audit event emitter and recording event emitter. (I believe this is what we had before these changes, although our naming/lack-of-documentation made it a bit difficult to track).

Perhaps related observation: The `s.cfg.AuthClient` that we return in `sync` mode is the same `*auth.Client` as the `s.cfg.Emitter` that we were passing to the `TeeStreamer` previously, see [here](https://github.com/gravitational/teleport/blob/8571856/lib/service/desktop.go#L210-L215).

zmb3 replied:
> my understanding is that session recording events and audit events are going to the same place

This is an easy mistake to make due to the way we've set things up, but your understanding is incorrect. Audit events are written to Teleport's audit backend (typically DynamoDB) via `cfg.Emitter`. Session recordings are written to the sessions backend (typically S3) via the `StreamWriter`.

Despite the different destinations, events are recorded using the `EmitAuditEvent` function in both cases (I'm really not a fan of this approach as it causes confusion)

The `TeeStreamer` is attempting to be clever, and I don't think clever is good here. Essentially it exposes a single `EmitAuditEvent` entry point, and it sends every event it receives to the recording (AKA stream), and a subset of the events (based on the event type) to Teleport's audit log.

What we're doing in this first commit by removing the `TeeStreamer` is moving the logic for which desktop events belong in session recordings and which desktop events belong in the audit log out of a shared component and into the `WindowsService` which knows best.

ibeckermayer replied:
I see. So before these changes, when we had

```go
func (s *WindowsService) newStreamer(ctx context.Context, recConfig types.SessionRecordingConfig) (libevents.Streamer, error) {
	if services.IsRecordSync(recConfig.GetMode()) {
		s.cfg.Log.Debugf("using sync streamer (for mode %v)", recConfig.GetMode())
		return s.cfg.AuthClient, nil
	}
	s.cfg.Log.Debugf("using async streamer (for mode %v)", recConfig.GetMode())
	uploadDir := filepath.Join(s.cfg.DataDir, teleport.LogsDir, teleport.ComponentUpload,
		libevents.StreamingSessionsDir, apidefaults.Namespace)
	fileStreamer, err := filesessions.NewStreamer(uploadDir)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	return libevents.NewTeeStreamer(fileStreamer, s.cfg.Emitter), nil
}
```

In that case, in `sync` recording mode, we would just send all audit and recording events directly to the auth server. In `async` mode, we would send [all audit and recording events to the `fileStreamer`](https://github.com/gravitational/teleport/blob/9fa28188c75c44baf9344fed1f1b2738577519f5/lib/events/emitter.go#L498), and [all audit, but not recording, events to the `s.cfg.Emitter` (aka the auth server)](https://github.com/gravitational/teleport/blob/9fa28188c75c44baf9344fed1f1b2738577519f5/lib/events/emitter.go#L503-L508).

With your proposed changes, `sync` mode still works the same (everything is sent to auth), however in `async` mode, _only recording events_ will be sent to the `fileStreamer`, no longer audit events. Audit events are now only sent to the auth server in `async` mode.

zmb3 committed:
Remove the checking stream

AuditWriter already uses the same checkAndSetEventFields call
that CheckingStreamer does, so we were running this logic twice
for each emitted event.

Also add a benchmark for audit writer with a file backend.

zmb3 committed:
Use klauspost/compress over stdlib gzip

This is a drop-in replacement with the same API.
I included a benchmark which shows a 10% performance improvement.

There are a few other places where performance is less critical
that I left using the stdlib gzip package for now.
